internal/plans/planfile/tfplan.go

```diff
@@ -90,41 +90,90 @@ func readTfplan(r io.Reader) (*plans.Plan, error) {
+		configAddrProxy, diags := addrs.ParseCheckableStr(objKind, rawCRs.ConfigAddr)
+		if diags.HasErrors() {
+			return nil, diags.Err()
 		}
-		switch r := rawCR.Result.(type) {
-		case *planproto.ConditionResult_Value:
-			cr.Result = cty.BoolVal(r.Value)
-		case *planproto.ConditionResult_Unknown:
-			cr.Result = cty.UnknownVal(cty.Bool)
+		configAddr := configAddrProxy.ConfigCheckable()
+		if configAddr.String() != configAddrProxy.String() {
+			// This is how we catch if the config address included index
+			// information that would be allowed in a Checkable but not
+			// in a ConfigCheckable.
+			return nil, fmt.Errorf("invalid checkable config address %s", rawCRs.ConfigAddr)
 		}
-		var diags tfdiags.Diagnostics
-		switch rawCR.Type {
-		case planproto.ConditionType_OUTPUT_PRECONDITION:
-			cr.Type = addrs.OutputPrecondition
-			cr.Address, diags = addrs.ParseAbsOutputValueStr(rawCR.Addr)
+
+		aggr.ObjectResults = addrs.MakeMap[addrs.Checkable, *states.CheckResultObject]()
+		for _, rawCR := range rawCRs.Objects {
+			objectAddr, diags := addrs.ParseCheckableStr(objKind, rawCR.ObjectAddr)
 			if diags.HasErrors() {
 				return nil, diags.Err()
 			}
-		case planproto.ConditionType_RESOURCE_PRECONDITION:
-			cr.Type = addrs.ResourcePrecondition
-			cr.Address, diags = addrs.ParseAbsResourceInstanceStr(rawCR.Addr)
-			if diags.HasErrors() {
-				return nil, diags.Err()
+			if !addrs.Equivalent(objectAddr.ConfigCheckable(), configAddr) {
+				return nil, fmt.Errorf("checkable object %s should not be grouped under %s", objectAddr, configAddr)
 			}
-		case planproto.ConditionType_RESOURCE_POSTCONDITION:
-			cr.Type = addrs.ResourcePostcondition
-			cr.Address, diags = addrs.ParseAbsResourceInstanceStr(rawCR.Addr)
-			if diags.HasErrors() {
-				return nil, diags.Err()
+
+			obj := &states.CheckResultObject{
+				FailureMessages: rawCR.FailureMessages,
 			}
-		default:
-			return nil, fmt.Errorf("condition result %s has unsupported type %s", rawCR.ConditionAddr, rawCR.Type)
+			switch rawCR.Status {
+			case planproto.CheckResults_UNKNOWN:
+				obj.Status = checks.StatusUnknown
+			case planproto.CheckResults_PASS:
+				obj.Status = checks.StatusPass
+			case planproto.CheckResults_FAIL:
+				obj.Status = checks.StatusFail
+			case planproto.CheckResults_ERROR:
+				obj.Status = checks.StatusError
+			default:
+				return nil, fmt.Errorf("object check results for %s has unsupported status %#v", rawCR.ObjectAddr, rawCR.Status)
+			}
+
+			aggr.ObjectResults.Put(objectAddr, obj)
 		}
-		plan.Conditions[conditionAddr] = cr
+		// If we ended up with no elements in the map then we'll just nil it,
+		// primarily just to make life easier for our round-trip tests.
+		if aggr.ObjectResults.Len() == 0 {
+			aggr.ObjectResults.Elems = nil
+		}
+
+		plan.Checks.ConfigResults.Put(configAddr, aggr)
+	}
+	// If we ended up with no elements in the map then we'll just nil it,
+	// primarily just to make life easier for our round-trip tests.
```

Comment:
Nilling these empty maps feels a bit tricky to me, only because elsewhere in this changeset we logically distinguish between an `addrs.Map`'s `Elems` being `nil` vs being an empty map (in `AggregateCheckStatus`). There's no bug here as far as I can tell, but I wonder if it would be clearer to be consistent by leaving this as an empty map.

(Noting also that the documentation for `addrs.Map` says that "callers MUST NOT modify this map directly", which also makes me think that this nilling is a little bit suspicious.)

Reply:
I believe what motivated this is that the file format for plans cannot actually distinguish between an empty map and a nil map -- we serialize as protobuf which has no explicit serialization for "map of zero elements" separate from there being no map set at all -- and so this extra special case is to pick one specific representation of "no results".

I must admit I don't recall if there was a reason why it was better for the canonical representation to be `nil` rather than being a zero-length non-nil map; I could try it and see if it's practical to change the _other_ end of this so that the "elems" are _never_ `nil` in this case. I think the main thing is that we need to decide one way or the other because we cannot possibly preserve the difference through protobuf serialization.